File: MessageDispatcher.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace socket_dispatcher_pseudo
{
    /// <summary>
    /// Dispatches incoming JSON messages to registered envelopes based on message name.
    /// </summary>
    public class MessageDispatcher
    {
        private readonly Dictionary<string, IMessageEnvelope> _envelopes = new();

        /// <summary>
        /// Registers a callback for a given message name.
        /// </summary>
        /// <param name="messageName">The logical name of the message type.</param>
        /// <param name="onReceived">The action to invoke when that message is received.</param>
        public void Register(string messageName, Action<IMessage> onReceived)
        {
            var env = MessageEnvelope.Create(messageName, new Dictionary<string, object>(), onReceived);
            _envelopes[messageName] = env;
        }

        /// <summary>
        /// Parses raw JSON, locates the registered envelope, and invokes its callback.
        /// </summary>
        /// <param name="json">The incoming JSON string.</param>
        public void Dispatch(string json)
        {
            var envelope = MessageEnvelope.FromJson(json);
            if (_envelopes.TryGetValue(envelope.Name, out var registered))
            {
                envelope.OnReceived = registered.OnReceived;
                envelope.Invoke();
            }
        }
    }

}

File: socket_dispatcher_python/core/message_dispatcher.py
from typing import Dict

from .message_envelope import MessageEnvelope
from stream_message import StreamMessage


class MessageDispatcher:
    """Routes incoming JSON messages to the registered envelopes."""

    def __init__(self) -> None:
        self._envelopes: Dict[str, MessageEnvelope] = {}

    def register(self, envelope: MessageEnvelope) -> None:
        if envelope.name in self._envelopes:
            raise ValueError(f"Envelope for '{envelope.name}' already registered")
        self._envelopes[envelope.name] = envelope

    def dispatch(self, json_str: StreamMessage) -> None:
       if json_str.Command in self._envelopes:
            registered = self._envelopes[envelope.name]
            envelope.on_received = registered.on_received
            envelope.invoke()
